### PR TITLE
demo testing a struct that returns a struct

### PR DESCRIPTION
**Pull request by strideynet**

### CODE REVIEW

File: lib/teleterm/daemon/config.go [hunk 1]
```diff
@@ -25,10 +27,28 @@ import (
...
 )
 
+type (
+	cluster interface {
+		GetServers(ctx context.Context) ([]clusters.Server, error)
+	}
+	storage interface {
+		GetByURI(uri string) (cluster, error)
+	}
+)
```
Comment (codingllama) [inline lib/teleterm/daemon/config.go:37]: nit: de-group interface declaration: less indent, a bit more common as an idiom.

```suggestion
type cluster interface {
	GetServers(ctx context.Context) ([]clusters.Server, error)
}

type storage interface {
	GetByURI(uri string) (cluster, error)
}
```

File: lib/teleterm/daemon/config.go [hunk 2]
```diff
@@ -25,10 +27,28 @@ import (
...
+
+// This is used to wrap a real clusters.Storage so it meets our interface
+type StorageWrapper struct {
```
Comment (codingllama) [inline lib/teleterm/daemon/config.go:40]: I would ditch the wrapper - code usage is king here, and should be enough to guarantee conformance to the interfaces. If there are no uses, then a temporary `var _ storage = &MyType{}` is enough to assert conformance.

Your method signatures are _not_ an accident, they are purposeful.

File: lib/teleterm/daemon/config.go [hunk 3]
```diff
@@ -25,10 +27,28 @@ import (
...
+	*clusters.Storage
+}
+
+func (sw *StorageWrapper) GetByURI(uri string) (cluster, error) {
+	return sw.GetByURI(uri)
```
Comment (codingllama) [inline lib/teleterm/daemon/config.go:45]: Looks like an infinite recursion to me?

```suggestion
	return sw.Storage.GetByURI(uri)
```

File: lib/teleterm/daemon/daemon_test.go [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+import (
+	"context"
+	"testing"
+
+	"github.com/gravitational/teleport/lib/teleterm/clusters"
+)
+
+type mockCluster struct {
+	cluster
```
Comment (codingllama) [inline lib/teleterm/daemon/daemon_test.go:11]: I would ditch this too - being used as the interface is enough to guarantee conformance.

File: lib/teleterm/daemon/daemon_test.go [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+	servers []clusters.Server
+}
+
+func (mc *mockCluster) GetServers(ctx context.Context) ([]clusters.Server, error) {
+	return mc.servers, nil
+}
+
+type mockStorage struct {
+	storage
```
Comment (codingllama) [inline lib/teleterm/daemon/daemon_test.go:20]: Same.

File: lib/teleterm/daemon/config.go (hunk 1 above)
Comment (strideynet) [reply lib/teleterm/daemon/config.go:37]: Yah, sorry, my editor auto formatted here ahhaa.

**Comment by codingllama** [comment]
The "tiny interfaces" pattern looks good to me, it's a reasonable and Go-like way to enable testing.

File: lib/teleterm/daemon/daemon_test.go (hunk 4 above)
Comment (strideynet) [reply lib/teleterm/daemon/daemon_test.go:11]: This is to demonstrate how you would override a single method of the cluster interface, for the purpose of testing, without having to stub out the rest of them.

File: lib/teleterm/daemon/daemon_test.go (hunk 5 above)
Comment (strideynet) [reply lib/teleterm/daemon/daemon_test.go:20]: As above.

File: lib/teleterm/daemon/config.go [hunk 6]
```diff
@@ -25,10 +27,27 @@ import (
...
+
+func (sw *StorageWrapper) GetByURI(uri string) (cluster, error) {
+	return sw.Storage.GetByURI(uri)
+}
+
 // Config is the cluster service config
 type Config struct {
 	// Storage is a storage service that reads/writes to tsh profiles
-	Storage *clusters.Storage
+	Storage storage
```
Comment (zmb3) [inline lib/teleterm/daemon/config.go:50]: It is a bit awkward for an exported field in an exported struct type to have an unexported type. I would export `Storage` too.

File: lib/teleterm/daemon/daemon_test.go (hunk 4 above)
Comment (codingllama) [reply lib/teleterm/daemon/daemon_test.go:11]: Gotcha. Hopefully whatever is being tested doesn't rely on all 10+ methods and could benefit from a smaller interface, but the technique is definitely useful. Embedding a concrete type might also make sense, depending on how it all works.

**Pull request closed without merging**